### PR TITLE
Feature/apmw 54 edit special locations

### LINKED ISSUE #54: Add/Edit "special" locations
As a Facility Manager, I want to be able to add new locations, such as benches or bicycles stands.
I also want to be able to edit these locations, for example when a bench is moved or gets a new paint job ;)
I also want to be able to "share" access to these locations with my colleagues, so that they can also edit the locations.
(one location can be associated with n users)

### Acceptance
- [ ] Ability to add new locations on website
- [ ] Ability to edit locations on website
- [ ] Share access to locations with other users

### DoD
- [ ] Tests
- [ ] Documentation
- [ ] CI Build passing
- [ ] Peer Review
- [ ] Approval from PO

### PR DESCRIPTION
**Pull request by taminob**
closes #54 and #53

Under `/locations/new` you can create a new location of interest.
create-form

Under `/locations` you will find a list with created locations.
list

You can click on `show` or visit `/locations/[id]` to show the details-page for a location.
details-page

### CODE REVIEW

File: app/views/locations/_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      <div class="field" id="opening-time-field-<%= o.object.id %>">
+        <%= o.select 'day',
+                     options_for_select({ 'Monday' => 0,
+                                          'Tuesday' => 1,
+                                          'Wednesday' => 2,
+                                          'Thursday' => 3,
+                                          'Friday' => 4,
+                                          'Saturday' => 5,
+                                          'Sunday' => 6
+                                        }, o.object.day),
```
Comment (taminob) [inline app/views/locations/_form.html.erb:51]: Worth a method in OpeningTimes? Weird that the other way around is a method but this not

File: app/views/locations/_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      const addNewOpeningTime = (target) => {
+          $(`<div class="field" id="opening-time-field-new">
+           <select name="location[openingtimes_attributes][-1][day]" id="opening-time-field-new">
+                  <option value="0">Monday</option>
+                  <option value="1">Tuesday</option>
+                  <option value="2">Wednesday</option>
+                  <option value="3">Thursday</option>
+                  <option value="4">Friday</option>
+                  <option value="5">Saturday</option>
+                  <option value="6">Sunday</option></select>
```
Comment (taminob) [inline app/views/locations/_form.html.erb:84]: Possible to use method of OpeningTimes? Weird code duplication

Comment (taminob) [reply app/views/locations/_form.html.erb:84]: Same for people (and probably buildings, but UI for opening times is still missing?)

File: app/models/openingtime.rb [hunk 3]
```diff
@@ -14,6 +14,19 @@ class Openingtime < ApplicationRecord
...
+      1 => 'Tuesday',
+      2 => 'Wednesday',
+      3 => 'Thursday',
+      4 => 'Friday',
+      5 => 'Saturday',
+      6 => 'Sunday'
+    }
+    mapping[day]
+  end
+
```
Comment (taminob) [inline app/models/openingtime.rb:29]: just used in tests? then, it should be a spec helper (better: use it!)

File: app/models/location.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# the model representing a location of interest
+class Location < ApplicationRecord
+  include Timeable
+
+  validates :name, :location_longitude, :location_latitude, presence: true
```
Comment (taminob) [inline app/models/location.rb:5]: location_latitude and location_longitude are checked twice for presence
maybe better: solution like timable with e.g. "include Locatable" which generates this (but probably not in this PR). What do you think? Should we open a new issue?

File: app/views/locations/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<h1>New Location</h1>
+
+<%= render 'form', location: @location %>
```
Comment (taminob) [inline app/views/locations/new.html.erb:3]: Maybe pass in string for submit button, related to #99

File: config/routes.rb [hunk 6]
```diff
@@ -12,7 +13,7 @@
       registrations: 'users/registrations',
       omniauth_callbacks: 'users/omniauth_callbacks'
     }
-  resources :users, only: [:show]
```
Comment (taminob) [inline config/routes.rb:15]: Is this just a left-over you removed or removed by accident?

File: spec/features/locations/edit_spec.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    visit location_path(@location)
+    expect(page).not_to have_css "img[alt='#{filename}']"
+
+    visit edit_location_path(@location)
+    page.attach_file('location[location_photo]', 'app/assets/images/default-profile-picture.png')
+    page.find('input[type=submit][name=commit]').click
+
+    visit location_path(@location)
+    expect(page).to have_css "img[alt='#{filename}']"
+  end
```
Comment (taminob) [inline spec/features/locations/edit_spec.rb:37]: weird test, why attach and then detach immediately? Why not just hardcode filename as it is done 3 lines above
Otherwise, not very pretty to test other route (show) in edit_spec, is this the right location? (but in edit_spec fine with me, no better idea; isn't this rather an integration test?)

File: spec/features/people/edit_person_details_spec.rb [hunk 8]
```diff
@@ -135,8 +135,7 @@
   end
 
   it "includes a select field to change the day of an openingtime" do
-    @openingtime = create :openingtime
-    @person.openingtimes = [@openingtime]
+    @openingtime = FactoryBot.create(:openingtime, timeable: @person)
```
Comment (taminob) [inline spec/features/people/edit_person_details_spec.rb:138]: Why did you change this? Unrelated to locations?

File: spec/models/building_spec.rb [hunk 9]
```diff
@@ -1,11 +1,16 @@
 require 'rails_helper'
 
 RSpec.describe Building, type: :model do
-  it "has name, location_latitude, location_longitude" do
-    building = FactoryBot.create :building
+  building = FactoryBot.create :building
+  openingtime = FactoryBot.create(:openingtime, timeable: building)
```
Comment (taminob) [inline spec/models/building_spec.rb:5]: should be fixture. Again, unrelated to locations?

File: spec/models/location_spec.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+RSpec.describe Location, type: :model do
+  location = FactoryBot.create :location
+  openingtime = FactoryBot.create(:openingtime, timeable: location)
```
Comment (taminob) [inline spec/models/location_spec.rb:5]: fixture?

File: spec/views/locations/edit.html.erb_spec.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+      assert_select "input[name=?]", "person[phone_number]"
+
+      assert_select "input[name=?]", "person[first_name]"
+
+      assert_select "input[name=?]", "person[last_name]"
+
+      assert_select "input[name=?]", "person[email]"
+    end
+  end
+end
```
Comment (taminob) [inline spec/views/locations/edit.html.erb_spec.rb:22]: wat? Why does this test person? o_0

File: spec/views/locations/index.html.erb_spec.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  it "renders a list of people" do
+    render
+    assert_select "tr>td", text: "+4990909090".to_s, count: 2
+    assert_select "tr>td", text: "Herbert".to_s, count: 2
+    assert_select "tr>td", text: "Herbertson".to_s, count: 2
+    assert_select "tr>td", text: "herbert.herbertson@hpi.de".to_s, count: 2
+  end
+end
```
Comment (taminob) [inline spec/views/locations/index.html.erb_spec.rb:15]: same here? People?

File: spec/views/locations/new.html.erb_spec.rb [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+      assert_select "input[name=?]", "person[phone_number]"
+
+      assert_select "input[name=?]", "person[first_name]"
+
+      assert_select "input[name=?]", "person[last_name]"
+
+      assert_select "input[name=?]", "person[email]"
+    end
+  end
+end
```
Comment (taminob) [inline spec/views/locations/new.html.erb_spec.rb:27]: nope, still not people

File: spec/views/locations/show.html.erb_spec.rb [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  it "renders attributes in <p>" do
+    render
+    expect(rendered).to match(/\+4990909090/)
+    expect(rendered).to match(/Herbert/)
+    expect(rendered).to match(/Herbertson/)
+    expect(rendered).to match(/herbert.herbertson@hpi.de/)
+  end
+end
```
Comment (taminob) [inline spec/views/locations/show.html.erb_spec.rb:15]: still nope, person != location, I'd say - change my mind

File: spec/views/people/edit.html.erb_spec.rb [hunk 15]
```diff
@@ -1,22 +1,26 @@
...
+      assert_select "input[name=?]", "location[location_photo]"
 
-      assert_select "input[name=?]", "person[email]"
+      assert_select "input[name=?]", "location[location_latitude]"
+
+      assert_select "input[name=?]", "location[location_longitude]"
+
+      assert_select "input[name=?]", "location[location_longitude]"
     end
   end
```
Comment (taminob) [inline spec/views/people/edit.html.erb_spec.rb:25]: hm? not sure what I should say. person? Won't comment on the others, just swap

File: spec/views/people/index.html.erb_spec.rb [hunk 16]
```diff
@@ -1,15 +1,13 @@
...
+  it "renders a list of locations" do
     render
-    assert_select "tr>td", text: "+4990909090".to_s, count: 2
-    assert_select "tr>td", text: "Herbert".to_s, count: 2
-    assert_select "tr>td", text: "Herbertson".to_s, count: 2
-    assert_select "tr>td", text: "herbert.herbertson@hpi.de".to_s, count: 2
+    assert_select "tr>td", text: "cafe".to_s, count: 2
+    assert_select "tr>td>img[src=?]", "/assets/default-location-photo.png", count: 2
   end
 end
```
Comment (taminob) [inline spec/views/people/index.html.erb_spec.rb:13]: .

File: spec/views/people/new.html.erb_spec.rb [hunk 17]
```diff
@@ -1,27 +1,32 @@
...
-                      last_name: "MyString",
-                      email: "MyString"
-                    ))
+    @location = create :location
+    #    assign(:person, Person.new(
+    #                      phone_number: "MyString",
+    #                      first_name: "MyString",
+    #                      last_name: "MyString",
+    #                      email: "MyString"
+    #                    ))
```
Comment (taminob) [inline spec/views/people/new.html.erb_spec.rb:11]: Won't comment on the person/location stuff, but why comment?

File: spec/views/people/show.html.erb_spec.rb [hunk 18]
```diff
@@ -1,15 +1,13 @@
...
+  it "renders attributes" do
     render
-    expect(rendered).to match(/\+4990909090/)
-    expect(rendered).to match(/Herbert/)
-    expect(rendered).to match(/Herbertson/)
-    expect(rendered).to match(/herbert.herbertson@hpi.de/)
+    expect(rendered).to match(/cafe/)
+    expect(rendered).to match(/cafe-details/)
   end
 end
```
Comment (taminob) [inline spec/views/people/show.html.erb_spec.rb:13]: :)

**Review by taminob — COMMENTED** [review]
I opened the PR, can't approve/request changes. @bewee  will do this.
Also, we're not sure about the default image, is a bit weird as default location icon. Any idea @TheGrayStone  ?

File: app/models/openingtime.rb (hunk 3 above)
Comment (cambryx) [reply app/models/openingtime.rb:29]: We already use it in the show-view for locations :)

File: app/views/locations/_form.html.erb (hunk 1 above)
Comment (cambryx) [reply app/views/locations/_form.html.erb:51]: Good idea. Will do this.

File: app/views/locations/_form.html.erb (hunk 2 above)
Comment (cambryx) [reply app/views/locations/_form.html.erb:84]: We use the new number to string mapping method from the openingtime model now

File: app/models/location.rb (hunk 4 above)
Comment (cambryx) [reply app/models/location.rb:5]: we removed the duplicate checks but did not introduce a "include Locatable"

File: app/views/locations/new.html.erb (hunk 5 above)
Comment (cambryx) [reply app/views/locations/new.html.erb:3]: We don't understand the problem here.
The issue #99 exist for people but we can't reproduce it for locations.

File: config/routes.rb (hunk 6 above)
Comment (cambryx) [reply config/routes.rb:15]: This is a leftover.
There used to be a users controller only for the show action. It has become obsolete since #55.

File: spec/features/locations/edit_spec.rb (hunk 7 above)
Comment (cambryx) [reply spec/features/locations/edit_spec.rb:37]: The reason for attaching and detaching is, that the actual filename is generated when attaching (it's a long number and not just default-profile-picture). Therefore we have to attach it first to find out the generated filename.

If you have you have concrete ideas how to improve this test, we are open to them :)
Otherwise we will leave the test as is.

File: spec/models/building_spec.rb (hunk 9 above)
Comment (Max784) [reply spec/models/building_spec.rb:5]: not sure what "fixtrue" means, but I agree we shouldn't have changed that on this branch.

File: spec/features/people/edit_person_details_spec.rb (hunk 8 above)
Comment (cambryx) [reply spec/features/people/edit_person_details_spec.rb:138]: At some point we had modified the factory for openingtimes and probably changed this as well.
Since the factorybot is back to the state it was before, we will revert this as well (there is no reason for this change now).

File: spec/models/building_spec.rb (hunk 9 above)
Comment (cambryx) [reply spec/models/building_spec.rb:5]: We changed it to a fixture now.
It doesn't really belong to this branch maybe, but we will leave it here.

File: spec/models/location_spec.rb (hunk 10 above)
Comment (cambryx) [reply spec/models/location_spec.rb:5]: will change it to before each

**Review by Tratori — APPROVED** [review]
